src/unix/filesystem.h

```diff
@@ -0,0 +1,44 @@
+#define resolve_dir(__dirfd, __path) ({ \
+    tuple cwd; \
+    if (*(__path) == '/') cwd = filesystem_getroot(current->p->fs); \
+    else if (__dirfd == AT_FDCWD) cwd = current->p->cwd; \
+    else { \
+        file f = resolve_fd(current->p, __dirfd); \
+        if (!is_dir(f->n)) return set_syscall_error(current, ENOTDIR); \
+        cwd = f->n; \
+    } \
+    cwd; \
+})
+
+static inline boolean is_dir(tuple n)
```

Comment:
I'm going to move linktarget() and lookup_sym() to filesystem.h, the other filesystem-related definitions as of now cannot be moved because they are used by code (tfs.c and stage3.c) that doesn't include unix internals.